crates/bevy_ui/src/render/material_pipeline.rs

```diff
@@ -0,0 +1,659 @@
+    },
+    Extract, ExtractSchedule, Render, RenderApp, RenderSet,
+};
+use bevy_transform::prelude::GlobalTransform;
+use bevy_utils::{FloatOrd, HashMap, HashSet};
+
+use crate::{
+    CalculatedClip, Node, RenderUiSystem, TransparentUi, UiMaterial, UiMaterialKey, UiStack,
+    UiVertex, QUAD_INDICES, QUAD_VERTEX_POSITIONS,
+};
+
+/// Adds the necessary ECS resources and render logic to enable rendering entities using the given
+/// [`UiMaterial`] asset type (which includes [`UiMaterial`] types).
+pub struct UiMaterialPlugin<M: UiMaterial>(PhantomData<M>);
+
+impl<M: UiMaterial> Default for UiMaterialPlugin<M> {
+    fn default() -> Self {
+        Self(Default::default())
+    }
+}
+
+impl<M: UiMaterial> Plugin for UiMaterialPlugin<M>
+where
+    M::Data: PartialEq + Eq + Hash + Clone,
+{
+    fn build(&self, app: &mut bevy_app::App) {
+        app.add_asset::<M>()
+            .add_plugins(ExtractComponentPlugin::<Handle<M>>::extract_visible());
+
+        if let Ok(render_app) = app.get_sub_app_mut(RenderApp) {
+            render_app
+                .add_render_command::<TransparentUi, DrawUiMaterial<M>>()
+                .init_resource::<ExtractedUiMaterials<M>>()
+                .init_resource::<ExtractedUiMaterialNodes<M>>()
+                .init_resource::<RenderUiMaterials<M>>()
+                .init_resource::<UiMatMeta<M>>()
+                .init_resource::<SpecializedRenderPipelines<UiMaterialPipeline<M>>>()
+                .add_systems(
+                    ExtractSchedule,
+                    (
+                        extract_ui_materials::<M>,
+                        extract_material_uinodes::<M>.in_set(RenderUiSystem::ExtractNode),
+                    ),
+                )
+                .add_systems(
+                    Render,
+                    (
+                        prepare_ui_materials::<M>
+                            .in_set(RenderSet::Prepare)
+                            .after(PrepareAssetSet::PreAssetPrepare),
+                        prepare_uimaterial_nodes::<M>.in_set(RenderSet::Prepare),
+                        queue_ui_material_nodes::<M>.in_set(RenderSet::Queue),
+                    ),
+                );
+        }
+    }
+
+    fn finish(&self, app: &mut App) {
+        if let Ok(render_app) = app.get_sub_app_mut(RenderApp) {
+            render_app.init_resource::<UiMaterialPipeline<M>>();
+        }
+    }
+}
+
+#[derive(Resource)]
+pub struct UiMatMeta<M: UiMaterial> {
+    vertices: BufferVec<UiVertex>,
+    view_bind_group: Option<BindGroup>,
+    marker: PhantomData<M>,
+}
+
+impl<M: UiMaterial> Default for UiMatMeta<M> {
+    fn default() -> Self {
+        Self {
+            vertices: BufferVec::new(BufferUsages::VERTEX),
+            view_bind_group: Default::default(),
+            marker: PhantomData,
+        }
+    }
+}
+
+#[derive(Component)]
+pub struct MatUiBatch<M: UiMaterial> {
```

Comment:
Same with this, I'd go with:
```suggestion
pub struct UiMaterialBatch<M: UiMaterial> {
```